lib/dialects/postgres/query-generator.js

```diff
@@ -363,21 +363,79 @@ const QueryGenerator = {
+    }
+    const uniqueIndexes = indexes.filter(e => e.unique).map(e => e.name);
+    const uniqueCount = uniqueKeys.length + uniqueIndexes.length;
+    const fields = Object.keys(model.fieldAttributeMap);
+    const fieldMap = {};
+    fields.forEach(field => {
+      fieldMap[model.fieldAttributeMap[field]] = field;
+    });
+
+    if ((uniqueCount > 1 && (!options.conflict || (options.conflict && !options.conflict.constraint))) ||
+      semver.lt(this.sequelize.options.databaseVersion, '9.5.0')) {
+      options.onConflictFallback = true;
+      const primaryField = this.quoteIdentifier(model.primaryKeyField);
+
+      let insert = this.insertQuery(tableName, insertValues, model.rawAttributes, options);
+      let update = this.updateQuery(tableName, updateValues, where, options, model.rawAttributes);
+
+      insert = insert.replace('RETURNING *', `RETURNING ${primaryField} INTO primary_key`);
+      update = update.replace('RETURNING *', `RETURNING ${primaryField} INTO primary_key`);
 
-    let insert = this.insertQuery(tableName, insertValues, model.rawAttributes, options);
-    let update = this.updateQuery(tableName, updateValues, where, options, model.rawAttributes);
+      return this.exceptionFn(
+        'sequelize_upsert',
+        tableName,
+        'OUT created boolean, OUT primary_key text',
+        `${insert} created := true;`,
+        `${update}; created := false`
+      );
+    }
+
+    const getUpdateArray = (values, target) =>
+      Object.keys(values).map(key => fieldMap[key] || key).filter(value => target.indexOf(value) === -1);
+
+    options.isPgUpsert = true;
+    if (!options.conflict) {
+      options.conflict = {
+        target: uniqueKeys,
+        update: getUpdateArray(updateValues, uniqueKeys)
+      };
+    }
+
+    if (options.conflict.update === false) {
+      options.onConflict = 'DO NOTHING';
+    } else {
+      options.conflict.target = options.conflict.target || uniqueKeys;
+      if (!Array.isArray(options.conflict.target)) {
+        options.conflict.target = [options.conflict.target];
+      }
+      options.conflict.target = options.conflict.target.map(key => fieldMap[key] || key);
+      if (!options.conflict.update) {
+        options.conflict.update = getUpdateArray(updateValues, options.conflict.target);
+      }
+
+      if (rawAttributes.updatedAt) {
+        insertValues.updatedAt = new Date();
+        if (options.conflict.update.indexOf('updatedAt') === -1) {
+          options.conflict.update.push('updatedAt');
+        }
+      }
+
+      options.onConflict = 'DO UPDATE SET ' + options.conflict.update.map(key => {
+        key = this.quoteIdentifier(key);
+        return key + ' = EXCLUDED.' + key;
+      }).join(', ');
+    }
 
-    insert = insert.replace('RETURNING *', `RETURNING ${primaryField} INTO primary_key`);
-    update = update.replace('RETURNING *', `RETURNING ${primaryField} INTO primary_key`);
+    options.returning = '*, (xmax = 0)::bool AS xmax';
```

Comment:
can you explain this xmax stuff a bit?

Reply:
Which postgres version you are using? I think it's supported since version 7.
https://www.postgresql.org/docs/current/static/ddl-system-columns.html
`xmax: The identity (transaction ID) of the deleting transaction, or zero for an undeleted row version`
i.e. `xmax = 0` means it is new record

Reply:
@inDream Latest - 10.1

Reply:
It passed the upsert test with version 10.2, could you give me the failed SQL?

Reply:
@inDream If you look at the stack, the error isn't from SQL - the error is in JS. It's trying to read a property of an object that doesn't exist. I'm able to reproduce it consistently when merging this branch into master, so it might be a conflict between this change and something else that landed on master.

Reply:
I've rebased you could try it again.